traefik.go

```diff
@@ -209,29 +209,36 @@ func main() {
 	log.Info("Shutting down")
 }
 
-func createTLSConfig(certFile string, keyFile string) (*tls.Config, error) {
+// creates a TLS config that allows terminating HTTPS for multiple domains using SNI
+func createTLSConfig(certs []Certificate) (*tls.Config, error) {
 	config := &tls.Config{}
 	if config.NextProtos == nil {
 		config.NextProtos = []string{"http/1.1"}
 	}
 
 	var err error
-	config.Certificates = make([]tls.Certificate, 1)
-	if len(certFile) > 0 && len(keyFile) > 0 {
-		config.Certificates[0], err = tls.LoadX509KeyPair(certFile, keyFile)
-		if err != nil {
-			return nil, err
+	if len(certs) > 0 {
+		config.Certificates = make([]tls.Certificate, len(certs))
+		for i, v := range certs {
+			config.Certificates[i], err = tls.LoadX509KeyPair(v.CertFile, v.KeyFile)
+			if err != nil {
+				return nil, err
+			}
 		}
+		// BuildNameToCertificate parses the CommonName and SubjectAlternateName fields
+		// in each certificate and populates the config.NameToCertificate map.
+		config.BuildNameToCertificate()
 	} else {
 		return nil, nil
 	}
+
 	return config, nil
 }
 
 func startServer(srv *manners.GracefulServer, globalConfiguration *GlobalConfiguration) {
 	log.Info("Starting server")
-	if len(globalConfiguration.CertFile) > 0 && len(globalConfiguration.KeyFile) > 0 {
-		err := srv.ListenAndServeTLS(globalConfiguration.CertFile, globalConfiguration.KeyFile)
+	if len(globalConfiguration.Certificates) > 0 {
+		err := srv.ListenAndServeTLS("", "")
```

Comment:
@emilevauge Mhhh. So Go 1.5 `ListenAndServeTLS` uses the TLSConfig from the srv struct (which we already populate in `prepareServer()`) by default.
> Filenames containing a certificate and matching private key for the server must be provided if the Server's TLSConfig.Certificates is not populated
https://golang.org/pkg/net/http/#Server.ListenAndServeTLS

But apparently mailgun/manners library hasn't been updated to provide the equivalent behavior.
But they have a `ListenAndServeTLSWithConfig` method which we can use.     https://github.com/mailgun/manners/blob/master/server.go#L174

I think we can do it this way:

```GO
func startServer(srv *manners.GracefulServer, globalConfiguration *GlobalConfiguration) {
	log.Info("Starting server")
	if srv.TLSConfig != nil {
		err := srv.ListenAndServeTLSWithConfig(srv.TLSConfig)
		if err != nil {
			log.Fatal("Error creating server: ", err)
		}
	} else {
```

And in `prepareServer()` set the TLSConfig in the server struct passed to HijackListener:
```GO
	server, err := oldServer.HijackListener(&http.Server{
		Addr:    globalConfiguration.Port,
		Handler: negroni,
		TLSConfig: tlsConfig,
	}, tlsConfig)
```

Good?